src/mqtt_client.rs

```diff
@@ -91,27 +99,29 @@ where
     ///
     /// # Returns
     /// An `MqttClient` that can be used for publishing messages and subscribing to topics.
-    pub fn new(broker: IpAddr, client_id: &str, network_stack: N) -> Result<Self, Error<N::Error>> {
-        // Connect to the broker's TCP port.
-        let socket = network_stack.open(Mode::NonBlocking)?;
-
-        // Next, connect to the broker over MQTT.
-        let socket = network_stack.connect(socket, SocketAddr::new(broker, 1883))?;
-
+    pub fn new(
+        broker: IpAddr,
+        client_id: &str,
+        network_stack: N,
+        clock: C,
+    ) -> Result<Self, Error<N::Error>> {
         let session_state = SessionState::new(
             broker,
             String::from_str(client_id).or(Err(Error::ProvidedClientIdTooLong))?,
         );
 
         let mut client = MqttClient {
             network_stack: network_stack,
-            socket: RefCell::new(Some(socket)),
+            socket: RefCell::new(None),
             state: RefCell::new(session_state),
             transmit_buffer: RefCell::new(GenericArray::default()),
             packet_reader: PacketReader::new(),
             connect_sent: false,
+            clock: clock,
         };
 
+        client.open_socket()?;
+        client.connect_socket()?;
```

Comment:
No socket operations are allowed during `new()`. If the network stack is not initialized (e.g. DHCP has not yet assigned an IP), we cannot get a socket yet and we would not be allowed to construct the client.

Reply:
Yep, this replicates the behaviour of current master, i.e. without #35.